File: scripts/CombineTable.pl
while (<>) {
@F=split/\s+/;

if (scalar(@ARGV) eq 3) {$PLUS{$F[0]}=$_ if $F[3]-$F[2] >= 200 and (!$PLUS{$F[0]} or $PLUS{$F[0]}=~m/\_/g)}
if (scalar(@ARGV) eq 2) {$MINUS{$F[0]}=$_ if $F[3]-$F[2] >= 200 and (!$MINUS{$F[0]} or $MINUS{$F[0]}=~m/\_/g)}
if (scalar(@ARGV) eq 1) {@G=split/\_/,$F[0];$g=$G[0]."\_".$G[1]."\_".$G[2];$INTACT{$g}=$_}
if (scalar(@ARGV) eq 0) {
$m=$F[1]-1999;
$p=$F[2]+2000;
$px=$F[2]+1;
$mx=$F[1]+1;
$mkey=$F[0]."\:".$m."\-".$F[1];
$pkey=$F[0]."\:".$px."\-".$p;
$ikey=$F[0]."\_".$mx."\_".$F[2];
@M=split/\t/,$MINUS{$mkey};
@P=split/\t/,$PLUS{$pkey};
@I=split/\t/,$INTACT{$ikey};
$intact="present";
$intact="intact" if $INTACT{$ikey};
$chr="NA";
$chr=$M[5] if $M[5] eq $P[5];
$strand="NA";
$strand="\+" if $M[4] eq $P[4] and $M[4] eq $F[5];
$strand="\-" if $M[4] eq $P[4] and $M[4] ne $F[5];
$start="NA";
$start=$M[8] if $M[7] <= $P[8] and $M[5] eq $P[5];
$start=$P[8] if $P[7] <= $M[8] and $M[5] eq $P[5];
$end="NA";
$end=$P[7] if $M[7] <= $P[8] and $M[5] eq $P[5];
$end=$M[7] if $P[7] <= $M[8] and $M[5] eq $P[5];
if ($end < $start) {$swap=$end; $end=$start; $start=$swap}
print "$F[0]\_$F[1]\_$F[2]\_$F[5]\_$F[4]\_$F[3]\_$intact\t$chr\_$start\_$end\_$strand\n";
}
}

File: haplongliner/combine_table.py
import sys
import re
from typing import Dict


def _read_minimap(file_path: str) -> Dict[str, str]:
    result = {}
    with open(file_path) as fh:
        for line in fh:
            if not line.strip():
                continue
            fields = line.strip().split()
            if len(fields) < 4:
                continue
            key = fields[0]
            aln_len = int(fields[3]) - int(fields[2])
            if aln_len >= 200 and (key not in result or "_" in result[key]):
                result[key] = line.strip()
    return result


def _read_intact(file_path: str) -> Dict[str, str]:
    result = {}
    with open(file_path) as fh:
        for line in fh:
            if not line.strip():
                continue
            fields = line.strip().split()
            m = re.match(r"^(.+?)_(\d+)_(\d+)_([+-])(?:_.*)?$", fields[0])
            if not m:
                continue
            chrom, start, end, _ = m.groups()
            key = f"{chrom}_{start}_{end}"
            result[key] = line.strip()
    return result


def combine_table(plus_file: str, minus_file: str, intact_file: str, fl_bed: str, out_file: str) -> None:
    plus = _read_minimap(plus_file)
    minus = _read_minimap(minus_file)
    intact = _read_intact(intact_file)

    with open(fl_bed) as bed, open(out_file, "w") as out:
        for line in bed:
            if not line.strip() or line.startswith("#"):
                continue
            f = line.strip().split()
            if len(f) < 6:
                continue
            chrom, start, end, name, dot, strand = f[:6]
            start_i = int(start)
            end_i = int(end)
            # Convert to the coordinate system used by the legacy pipeline.
            m = start_i - 1999
            p = end_i + 2000
            px = end_i + 1
            mx = start_i + 1
            mkey = f"{chrom}:{m}-{start_i}"
            pkey = f"{chrom}:{px}-{p}"
            # ORF headers store 1-based coordinates
            ikey = f"{chrom}_{mx}_{end_i}"

            m_val = minus.get(mkey, "").split("\t") if minus.get(mkey) else []
            p_val = plus.get(pkey, "").split("\t") if plus.get(pkey) else []

            status = "present"
            if ikey in intact:
                status = "intact"

            chr_ref = "NA"
            if len(m_val) > 5 and len(p_val) > 5 and m_val[5] == p_val[5]:
                chr_ref = m_val[5]

            out_strand = "NA"
            if len(m_val) > 4 and len(p_val) > 4:
                if m_val[4] == p_val[4] and m_val[4] == strand:
                    out_strand = "+"
                elif m_val[4] == p_val[4] and m_val[4] != strand:
                    out_strand = "-"

            start_ref = "NA"
            end_ref = "NA"
            if len(m_val) > 8 and len(p_val) > 8 and len(m_val) > 8 and m_val[5] == p_val[5]:
                if int(m_val[7]) <= int(p_val[8]):
                    start_ref = m_val[8]
                    end_ref = p_val[7]
                if int(p_val[7]) <= int(m_val[8]):
                    start_ref = p_val[8]
                    end_ref = m_val[7]
                if start_ref != "NA" and end_ref != "NA" and int(end_ref) < int(start_ref):
                    start_ref, end_ref = end_ref, start_ref

            out.write(
                f"{chrom}_{start}_{end}_{strand}_{dot}_{name}_{status}\t{chr_ref}_{start_ref}_{end_ref}_{out_strand}\n"
            )


if __name__ == "__main__":
    if len(sys.argv) != 6:
        sys.exit(
            "Usage: python -m haplongliner.combine_table <plus> <minus> <intact> <bed> <out>"
        )
    combine_table(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
